hydra/core/callbacks.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import warnings
+from typing import Any, List
+
+from omegaconf import DictConfig
+
+from hydra._internal.callback import Callback
+from hydra.core.singleton import Singleton
+from hydra.core.utils import JobReturn
+from hydra.utils import instantiate
+
+
+class Callbacks(metaclass=Singleton):
+    @staticmethod
+    def instance(*args: Any, **kwargs: Any) -> "Callbacks":
+        ret = Singleton.instance(Callbacks, *args, **kwargs)
+        assert isinstance(ret, Callbacks)
+        return ret
+
+    def __init__(self) -> None:
+        self.callbacks: List[Callback] = []
+
+    def instantiate_callbacks(self, config: DictConfig) -> None:
+        if config.hydra.callbacks is None:
+            raise RuntimeError("Hydra Callbacks is not configured")
+
+        assert len(self.callbacks) == 0, "Callbacks already instantiated."
+
+        for params in config.hydra.callbacks.values():
+            self.callbacks.append(instantiate(params))
+
+    def _iterate(
+        self,
+        function_name: str,
+        **kwargs: Any,
+    ) -> None:
+        for c in self.callbacks:
+            try:
+                getattr(c, function_name)(**kwargs)
+            except Exception as e:
+                warnings.warn(
+                    f"Exception occurred during {function_name} with callback {c}, error message: {e}"
+                )
```

Comment:
What does the exception message looks like? in particular, the callback {c}?

Reply:
I've fixed the warning message a bit, looks like the following now.
```
[2021-03-23 16:52:57,503][HYDRA] Init DefaultCallback with self.param1=foo, self.param2=bar
/Users/jieru/workspace/hydra-fork/hydra/hydra/core/callbacks.py:42: UserWarning: Exception occurred during on_run_start with callback DefaultCallback, error message: hello world
  warnings.warn(
```